### ISSUE TITLE
Service network-addresses-<iface>.service fails if an address is already assigned to an interface.

### ISSUE THREAD

**Issue by tom-kuca** [NixOS/nixpkgs#306841]
### Describe the bug
Service `network-addresses-<iface>.service` fails if an address is already assigned to an interface.

```
systemd[1]: Starting Address configuration of venet0...
network-addresses-venet0-start[48854]: adding address 2a03:3b40:101:6f:3ca6:ef1a:e811:5488/64... 'ip addr add 2a03:3b40:101:6f:3ca6:ef1a:e811:5488/64 dev venet0' failed: Error: ipv6: address already assigned.
systemd[1]: network-addresses-venet0.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: network-addresses-venet0.service: Failed with result 'exit-code'.
```

 

### Steps To Reproduce

1. Add an IP address to an interface
```
 ip addr add  "2a03:3b40:101:6f::b00:dead" dev venet0
 ```

2. Add the same IP address to the nixos configuration
```nix
 networking.interfaces.venet0.ipv6.addresses = ["2a03:3b40:101:6f:3ca6:ef1a:e811:5488"];
 ```
3. Switch to the new configuration and start the service
```
systemctl start network-addresses-venet0.service
```

### Expected behavior

The service should skip addresses that are already assigned and finish with success.

### Additional context

There is a check in the script run by the service (generated [here](https://github.com/NixOS/nixpkgs/blob/bbe2490a512b66964b48f6b431897dbed8164894/nixos/modules/tasks/network-interfaces-scripted.nix#L208)):
```bash
echo "2a03:3b40:101:6f:3ca6:ef1a:e811:5488/64" >> $state
echo -n "adding address 2a03:3b40:101:6f:3ca6:ef1a:e811:5488/64... "
if out=$(ip addr add "2a03:3b40:101:6f:3ca6:ef1a:e811:5488/64" dev "venet0" 2>&1); then
  echo "done"
elif ! echo "$out" | grep "File exists" >/dev/null 2>&1; then
  echo "'ip addr add "2a03:3b40:101:6f:3ca6:ef1a:e811:5488/64" dev "venet0"' failed: $out"
  exit 1
fi
```

It checks presence of _File exists_ in the output. This likely matches the message `RTNETLINK answers: File exists`, that was there sometimes earlier[¹](https://unix.stackexchange.com/questions/100588/using-ip-addr-instead-of-ifconfig-reports-rtnetlink-answers-file-exists-on-de). Now the error generated by `ip addr add` is different:

```console
> sudo ip addr add  "2a03:3b40:101:6f::b00:dead" dev venet0
Error: ipv6: address already assigned.
```

I got this error message both on 20.09 and nixos-unstable.

If I replace _File exists_ with _address already assigned_ in the script, it works as expected.

Note: If IP addresses are managed only by nixos, it works correctly. They are deleted when service is stopped. The problem is only present when there is an address that was added manually or by non-nixos script.

### Notify maintainers

@rnhmjoj

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.28, NixOS, 24.05 (Uakari), 24.05.20240410.1042fd8`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.21.2`
 - channels(root): `"nixos-23.11pre494976.7c67f006ea0"`
 - nixpkgs: `/nix/store/r1g09wfd968168f6drdl7agfaf3msik4-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by rnhmjoj** [comment]
> It checks presence of File exists in the output. This likely matches the message RTNETLINK answers: File exists, that was there sometimes earlier[¹](https://unix.stackexchange.com/questions/100588/using-ip-addr-instead-of-ifconfig-reports-rtnetlink-answers-file-exists-on-de). Now the error generated by ip addr add is different:

Ah, great detective work!

Perhaps it would be better to run `ip addr del <address> || true` and then `ip addr add <address>`.
Relying on the error message literal text is not a good idea. What do you think?

**Comment by tom-kuca** [comment]
I also don't like relying on the error message. 

There is `ip addr replace`, that looks better than add+del. No info in man page, better description is on [serverfault](https://serverfault.com/questions/476926/understanding-ip-addr-change-and-ip-addr-replace-commands).

I see one downside - if there is additional configuration associated with IP address, it will likely remove it. More complex solution is to pre-check if address is already assigned, something like 
```
ip -brief  addr show to "2a03:3b40:101:6f::b00:dead/64"  dev venet0 | grep 2a03:3b40:101:6f::b00:dead/64
```

I'm not sure if there are some corner cases with this check, so I would probably stick with less complex `ip addr replace`.

**Comment by martinholters** [comment]
Has this regressed in some way in 24.05? I have extra addresses configured on `lo` and it now fails due this issue because 127.0.0.1 seems to be configured before `network-addresses-lo.service` runs. But it did work before I updated to 24.05.

**Comment by mattchrist** [comment]
I'm also experiencing this after updating to 24.05.

On some of my systems I'm setting kernel `ip=` options to setup an interface initially (for remote unlocking of zfs datasets) and since updating to 24.05 the interface service fails to start because this address is already there.

**Comment by rnhmjoj** [comment]
Does https://github.com/NixOS/nixpkgs/pull/312317 fixes the issue for you?

**Comment by martinholters** [comment]
I've monkey-patched the change from #312317 into my configuration and that fixes it, yes.

**Comment by rnhmjoj** [comment]
You can track the progress here for [24.05](https://nixpk.gs/pr-tracker.html?pr=316911) and [unstable](https://nixpk.gs/pr-tracker.html?pr=312317).
